### PR TITLE
Feature: switch visibility with update_repo_settings #2537

### PR DESCRIPTION
**Pull request by WizKnight**
This PR enhances the `update_repo_settings` method to provide more control over repository settings, and deprecates the `update_repo_visibility` method:

**Key changes:**

* Added a private parameter to set repositories as public or private.
* Made the gated and private parameters optional, allowing updates to either setting independently.
* Added checks to ensure at least one of `gated` or `private` is provided.
* Payload Optimization: `gated` and `private` are only included in the request payload if they have non-None values.
* Updated the docstring for `update_repo_settings` and replaced all references to `update_repo_visibility` in the documentation.
* Added the `@_deprecate_method` decorator to `update_repo_visibility`, marking it for removal in v0.29.x.

### CODE REVIEW

**Comment by WizKnight** [comment]
Hey @Wauplin, the changes in this PR are implemented as per the TO-DOs in issue #2537.

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:3529]:
```suggestion
        version="0.29", message="Use `update_repo_settings` instead."
```
nit: we can specify a shorter message for conciseness

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:127]: `_deprecate_method` is already imported in [line 139](https://github.com/huggingface/huggingface_hub/blob/64bcff5c7725f29514b15d4db3a7a92e3955e577/src/huggingface_hub/hf_api.py#L139)

File: src/huggingface_hub/utils/__init__.py
Comment (hanouticelina) [inline src/huggingface_hub/utils/__init__.py:49]: No need to import `_deprecate_method`, see the comment Line 127 in `huggingface_hub/hf_api.py`

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:3588]: I think the default values for both `gated` and `private` attributes should be set to `None` instead

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:3628]: We need to raise a `ValueError` if both `gated` and `private` are None

**Review by hanouticelina — CHANGES_REQUESTED** [review]
Hi @WizKnight! thank you for working on this 🤗 I left some comments and I think they are still other references of `update_repo_visibility` in other files and docs to update

File: src/huggingface_hub/utils/__init__.py
Comment (WizKnight) [reply src/huggingface_hub/utils/__init__.py:49]: I had to import it, otherwise the test runs into an `ImportError: cannot import name '_deprecate_method' from 'huggingface_hub.utils'`.

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [reply src/huggingface_hub/hf_api.py:127]: oh okay, that's weird that this line comes from merging main to your branch, in the current version of main, `_deprecate_method` is imported [line 139](https://github.com/huggingface/huggingface_hub/blob/64bcff5c7725f29514b15d4db3a7a92e3955e577/src/huggingface_hub/hf_api.py#L139C1-L139C50).
So, I think line 127 should be removed and the import in [`huggingface_hub/utils/__init__.py`](https://github.com/huggingface/huggingface_hub/blob/5e3aef1bda0a846e045a8144393bd0b1ffe98f56/src/huggingface_hub/utils/__init__.py#L49) as well

Comment (WizKnight) [reply src/huggingface_hub/hf_api.py:127]: > oh okay, that's weird that this line comes from merging main to your branch, in the current version of main, `_deprecate_method` is imported [line 139](https://github.com/huggingface/huggingface_hub/blob/64bcff5c7725f29514b15d4db3a7a92e3955e577/src/huggingface_hub/hf_api.py#L139C1-L139C50). So, I think line 127 should be removed and the import in [`huggingface_hub/utils/__init__.py`](https://github.com/huggingface/huggingface_hub/blob/5e3aef1bda0a846e045a8144393bd0b1ffe98f56/src/huggingface_hub/utils/__init__.py#L49) as well

Alright then, I'll do that.

**Comment by WizKnight** [comment]
Hey @hanouticelina :), changes were added in the recent commit.

**Review by hanouticelina — CHANGES_REQUESTED** [review]
Hi @WizKnight! thanks a lot for this iteration! 😄  I left a small comment and there is also [this one](https://github.com/huggingface/huggingface_hub/pull/2541/files#r1766514200) that needs to be addressed. 
I just realized that `update_repo_visibility()` is used in other tests in `tests/test_file_download.py` and `tests/test_snapshot_download.py`. Could you replace `update_repo_visibility()` with `update_repo_settings()` in those files as well?
The failing test seems unrelated, I'll take a look

**Comment by Wauplin** [comment]
> I just realized that update_repo_visibility() is used in other tests in tests/test_file_download.py and tests/test_snapshot_download.py. Could you replace update_repo_visibility() with update_repo_settings() in those files as well?

For deprecated methods, we have a `expect_deprecation` decorator that can be used in tests (see [here](https://github.com/huggingface/huggingface_hub/blob/64bcff5c7725f29514b15d4db3a7a92e3955e577/tests/test_file_download.py#L190)). The goal is to continue to have the method tested since it's still officially supported and we are also testing that the method is correctly deprecated (i.e. correctly triggers warning).

So in the case of the deprecate use of it [here](https://github.com/huggingface/huggingface_hub/blob/64bcff5c7725f29514b15d4db3a7a92e3955e577/tests/test_snapshot_download.py#L98), you can do

```py
    @expect_deprecation("update_repo_visibility")
    def test_download_private_model(self):
        self.api.update_repo_visibility(repo_id=self.repo_id, private=True)
```
(+ adapt the import)

(sorry it's the kind of undocumentated things...)

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:215]: For tests that are using `update_repo_visibility`, it's best to keep them as they are now and use the `@expect_deprecation` decorator (see comment above). The philosophy behind it is that we shouldn't change existing behavior for end users so we shouldn't change the tests. 

It's only when we will definitely remove the method from the library that we will modify the tests.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3527]:
```suggestion
    @_deprecate_method(version="0.29", message="Please use `update_repo_settings` instead.")
    @validate_hf_hub_args
```

(still good to validate HF args even if deprecated)

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3536]:
```suggestion
        """Update the visibility setting of a repository.
        
        Deprecated. Use `update_repo_settings` instead.
```

Better to put it in a second line so that we can recommend to the user what they should use

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3602]: Should remove `* False (default): The ...` and replace by `* False: The`.

And above in the docstring: 

```
The gated status for the repository. If set to `None` (default), the `gated` setting of the repository won't be updated
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3637]: I would prepare the payload in 2 steps:

```py
payload: Dict = {}

if gated is not None:
    if gated not in ["auto", "manual", False]:
        raise ValueError(f"Invalid gated status, must be one of 'auto', 'manual', or False. Got '{gated}'.")
    payload["gated"] = gated

if private is not None:
    payload["private"] = private
```

The advantage of doing so is that adding a new parameter will just be a matter of adding a new `if` statement. Also, it makes it easier to log/print the payload value in case of debugging an issue. Doing a `{key: value for key, value in payload.items() if value is not None}` is not wrong per-se but makes it slightly less readable and less debuggable

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:163]: I would structure the documentation as such:

```md

## Change repository settings

(...) current content

### Update visibility

(...) current content + update example

### Setup gated access

(...) current content
```

Setting a repo as gating and setting repo visibility is 2 different topics, even if related to accessibility. Usually either a repo is public, either private, either public and gated. Setting a repo as private and gated is possible but doesn't make sense.

**Review by Wauplin — COMMENTED** [review]
Thanks @WizKnight! Here are a few changes I'd make in addition to @hanouticelina's comment above :)

File: src/huggingface_hub/hf_api.py
Comment (WizKnight) [reply src/huggingface_hub/hf_api.py:3637]: I was planning on a similar approach, but I opted for the dictionary comprehension `{key: value for key, value in payload.items() if value is not None}` to make the code more efficient.

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:3602]:
```suggestion
                The gated status for the repository. If set to `None` (default), the `gated` setting of the repository won't be updated
```

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:3605]:
```suggestion
                * False : The repository is not gated, and anyone can access it.
```

**Review by hanouticelina — CHANGES_REQUESTED** [review]
Hi @WizKnight ! thanks again for this 😄 I left two minor comments.

The tests failure comes from the test `test_repo_id_no_warning()` which is flagged with the `expect_deprecation` decorator and contains also a context manager `warnings.catch_warnings` and they both try to capture the same warning. 
The `FutureWarning` is "consumed" by `pytest.warns` in the `expect_deprecation` decorator, so it never reaches the `warnings.catch_warnings` context manager inside the test. This is why the `record` list does not contain the `FutureWarning` and the test fails. 

@Wauplin do we still need this test ? it tests that there is no warnings when passing `repo_id` as positional argument into some HfApi functions (`create_repo`, `delete_repo`, and `update_repo_visibility`).

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: Sometime readibility / maintenance is more important than efficiency :)

**Comment by Wauplin** [comment]
> @Wauplin do we still need this test ? it tests that there is no warnings when passing repo_id as positional argument into some HfApi functions (create_repo, delete_repo, and update_repo_visibility).

Clearly not needed anymore indeed! Let's just delete it instead of trying to fix the warning capture thingy (we could extend `expect_deprecation` to handle multiple inputs but really not worth it)

**Comment by WizKnight** [comment]
> HfApi functions (`create_repo`, `delete_repo`, and `update_repo_visibility`).

Hey @Wauplin and @hanouticelina, so do I remove these HfApi functions (`create_repo`, `delete_repo`, and `update_repo_visibility`) completely or extend those functions with `expect_deprecation` ??

**Comment by Wauplin** [comment]
No no, the only thing to do is to remove the test `test_repo_id_no_warning`. It's an outdated and legacy one. The methods create_repo and delete_repo are far from been deprecated!

**Comment by WizKnight** [comment]
Hey @Wauplin I removed the test `test_repo_id_no_warning` completely and iterated on @hanouticelina minor comments in the recent commit.

**Comment by Wauplin** [comment]
(I've switched the PR as "ready for review". In general @WizKnight you can switch a PR to "ready" as soon as you request someone to review it)

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:149]:
```suggestion
### Update visibility
```

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:157]:
```suggestion
>>> from huggingface_hub import update_repo_settings
>>> update_repo_settings(repo_id=repo_id, private=True)
```

Let's keep the initial pattern (i.e. not importing HfApi)

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:162]:
```suggestion
To give more control over how repos are used, the Hub allows repo authors to enable **access requests** for their repos. User must agree to share their contact information (username and email address) with the repo authors to access the files when enabled. A repo with access requests enabled is called a **gated repo**.
```

why this change? :smile:

**Review by hanouticelina — APPROVED** [review]
Hey @WizKnight 👋 thanks a lot for this PR! looks good to me 🙂 failing tests are unrelated, let's wait for @Wauplin's final review and we will be good to merge

**Review by Wauplin — APPROVED** [review]
Looking good! Thanks for the back and forth @WizKnight @hanouticelina :hugs: 

Failing test is indeed unrelated. I left minor comments otherwise we are good to merge :)

**Comment by Wauplin** [comment]
CI's green, let's merge :tada: Thanks again @WizKnight :hugs:

**Comment by WizKnight** [comment]
Hey @Wauplin & @hanouticelina , I really appreciate the opportunity!! Thankyou once again🤗

**Pull request merged**
